File: pages/refactorings/extract-method/01-order-total/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("total: Before == After", totalAgrees),
    property("no discount, no tax: total == subtotal", plainTotalIsSubtotal),
  )

  // Raw values, so the same input can be fed to both Before.Order and After.Order.
  val genLine: Gen[(Int, Int)] =
    for p <- Gen.int(Range.linear(-100, 1000)); q <- Gen.int(Range.linear(0, 20)) yield (p, q)
  val genOrder: Gen[(List[(Int, Int)], Int, Int)] =
    for
      items <- genLine.list(Range.linear(0, 10))
      d     <- Gen.int(Range.linear(0, 100))
      t     <- Gen.int(Range.linear(0, 30))
    yield (items, d, t)

  def totalAgrees: Property =
    for o <- genOrder.forAll
    yield
      val (items, d, t) = o
      Before.total(Before.Order(items.map(Before.Line(_, _)), d, t))
        ==== After.total(After.Order(items.map(After.Line(_, _)), d, t))

  def plainTotalIsSubtotal: Property =
    for items <- genLine.list(Range.linear(0, 10)).forAll
    yield
      val lines = items.map(After.Line(_, _))
      After.total(After.Order(lines, 0, 0)) ==== After.subtotal(lines)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default), t.result, Seed.fromTime())
    println(Test.renderReport("Props", t, r, ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)

File: pages/refactorings/extract-method/01-order-total/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

-- Raw values, so the same input can be fed to both Before.Order and After.Order.
genLine :: Gen (Int, Int)
genLine = (,) <$> Gen.int (Range.linear (-100) 1000) <*> Gen.int (Range.linear 0 20)

genOrder :: Gen ([(Int, Int)], Int, Int)
genOrder = (,,) <$> Gen.list (Range.linear 0 10) genLine
                <*> Gen.int (Range.linear 0 100)
                <*> Gen.int (Range.linear 0 30)

prop_total_agrees :: Property
prop_total_agrees = property $ do
  (items, d, t) <- forAll genOrder
  Before.total (Before.Order (map (uncurry Before.Line) items) d t)
    === After.total (After.Order (map (uncurry After.Line) items) d t)

prop_plain_total_is_subtotal :: Property
prop_plain_total_is_subtotal = property $ do
  items <- forAll (Gen.list (Range.linear 0 10) genLine)
  let ls = map (uncurry After.Line) items
  After.total (After.Order ls 0 0) === After.subtotal ls

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("total: Before == After", prop_total_agrees)
    , ("no discount, no tax: total == subtotal", prop_plain_total_is_subtotal)
    ]
  unless ok exitFailure
